OpenRA.Mods.Common/Traits/World/DomainIndex.cs

```diff
@@ -23,31 +23,32 @@ class DomainIndexInfo : TraitInfo<DomainIndex> { }
 
 	public class DomainIndex : IWorldLoaded
 	{
+		TileSet tileSet;
 		Dictionary<uint, MovementClassDomainIndex> domainIndexes;
 
 		public void WorldLoaded(World world, WorldRenderer wr)
 		{
 			domainIndexes = new Dictionary<uint, MovementClassDomainIndex>();
-			var tileSet = world.Map.Rules.TileSet;
-			var movementClasses =
-				world.Map.Rules.Actors.Where(ai => ai.Value.HasTraitInfo<MobileInfo>())
-					.Select(ai => (uint)ai.Value.TraitInfo<MobileInfo>().GetMovementClass(tileSet)).Distinct();
+			tileSet = world.Map.Rules.TileSet;
+			var locomotors = world.WorldActor.TraitsImplementing<Locomotor>().Where(l => !string.IsNullOrEmpty(l.Info.Name));
+			var movementClasses = locomotors.Select(t => (uint)t.Info.GetMovementClass(tileSet)).Distinct();
 
 			foreach (var mc in movementClasses)
 				domainIndexes[mc] = new MovementClassDomainIndex(world, mc);
 		}
 
-		public bool IsPassable(CPos p1, CPos p2, MobileInfo mi, uint movementClass)
+		public bool IsPassable(CPos p1, CPos p2, LocomotorInfo li)
 		{
 			// HACK: Work around units in other movement layers from being blocked
 			// when the point in the main layer is not pathable
 			if (p1.Layer != 0 || p2.Layer != 0)
 				return true;
 
 			// HACK: Workaround until we can generalize movement classes
-			if (mi.Subterranean || mi.Jumpjet)
+			if (li is SubterraneanLocomotorInfo || li is JumpjetLocomotorInfo)
```

Comment:
This PR has generalized movement classes enough to remove this hack:
* Add a `public virtual bool DisableDomainPassabilityCheck { get { return false; } }` property to `LocomotorInfo`
* Override it to return false in `SubterraneanLocomotorInfo` and `JumpjetLocomotorInfo`
* Change this to `if (li.DisableDomainPassabilityCheck)` and remove the `// HACK` comment.